### PR TITLE
chore(orchestration): endowments are membrane-friendly

### PR DESCRIPTION
**Pull request by 0xpatrickdev**
closes: #XXXX
refs: #9449
refs: #9308

## Description

Changes in this PR are related to #9449, primarily for `local-orchestration-account.js` and `cosmos-orchestration-account.js`, `service.js`, and `orchestrator.js`. Attenuated versions of these kit are what the caller of `...getChain('agoric' | 'cosmos').makeAccount()` receives. 

This PR also includes a change to `smart-wallet` so it understands how to unwrap **offerResult**s that are **Vow**s. This  uses **heapVowTools** which means it does  _**not**_  cover resumability in the scenario of `smart-wallet` upgrading - only the contracts that rely on it.

Other small changes included changes:
- `PromiseToVow` and `VowifyAll` type helpers, so we can still reference the idealized API spec in code that returns Vows
- fix/refactor of `getTimeoutTimestampNS` logic that removes an unnecessary network call when the caller supplies values
- lint rule: adds a warning when importing `heapVowE` in resumable code
- resumable lint rule changes to **error**, now that all warnings are resolved

### Security Considerations
N/A

### Scaling Considerations
N/A

### Documentation Considerations
N/A

### Testing Considerations

Adds a test boot and unit tests for a rejected Delegate wallet offer, which is helpful for ensuring errors are properly propagated through the vow chain.

Upgrade tests would be helpful here, but we haven't started on these yet. I don't think we need those to land this PR, although they are certainly critical before release. Would love to pair with someone on this and learn more about how we do that. 

### Upgrade Considerations

This PR contains a change to **smart-wallet**.

### CODE REVIEW

File: packages/boot/tools/supports.ts [hunk 1]
```diff
@@ -416,9 +416,21 @@ export const makeSwingsetTestKit = async (
...
             return 'agoric1mockVlocalchainAddress';
-          case 'VLOCALCHAIN_EXECUTE_TX':
-            // returns one empty object per message
-            return obj.messages.map(() => ({}));
+          case 'VLOCALCHAIN_EXECUTE_TX': {
+            return obj.messages.map(message => {
+              switch (message['@type']) {
+                case '/cosmos.staking.v1beta1.MsgDelegate': {
+                  if (message.amount.amount === '504') {
+                    throw Error('simulated packet timeout');
```
Comment (0xpatrickdev) [inline packages/boot/tools/supports.ts:426]: What's the best way to communicate a failure here? (bridge in bootstrap test)

File: packages/orchestration/src/exos/local-orchestration-account.js [hunk 2]
```diff
@@ -143,27 +147,28 @@ export const prepareLocalOrchestrationAccountKit = (
...
-          return zcf.makeInvitation(async seat => {
-            // TODO should it allow delegating more BLD?
+          return zcf.makeInvitation(seat => {
             seat.exit();
-            return this.facets.holder.delegate(validatorAddress, ertpAmount);
+            return watch(
+              this.facets.holder.delegate(validatorAddress, ertpAmount),
+            );
           }, 'Delegate');
         },
```
Comment (0xpatrickdev) [inline packages/orchestration/src/exos/local-orchestration-account.js:157]: What is the correct way to return an invitation here? I think we only need the offer handler to be resumable, which I've done here. This, or perhaps something else, seems to result in errors being swallowed.

File: packages/boot/test/bootstrapTests/lca.test.ts [hunk 3]
```diff
@@ -89,4 +89,30 @@ test.serial('stakeBld', async t => {
...
+          // @ts-expect-error XXX BoardRemote
+          In: { brand: BLD, value: 504n },
+        },
+      },
+    }),
+    {
+      // FIXME should receive "simulated packet timeout"
+      message:
+        /syscall.callNow failed: device.invoke failed, see logs for details/,
+    },
```
Comment (0xpatrickdev) [inline packages/boot/test/bootstrapTests/lca.test.ts]: This newly added test, despite the broken error propagation through runUtils, passes on current `master` (but is failing here)

Comment (0xpatrickdev) [reply packages/boot/test/bootstrapTests/lca.test.ts]: Temporarily fixed via [d77582b](https://github.com/Agoric/agoric-sdk/pull/9591/commits/d77582b77f2b5f683520a8a6eb0bf47202e7fe1a)

File: packages/orchestration/src/exos/local-orchestration-account.js (hunk 2 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/exos/local-orchestration-account.js:157]: This seems to be correct. The issue faced is due to https://github.com/Agoric/agoric-sdk/issues/9308

File: packages/orchestration/src/exos/cosmos-orchestration-account.js [hunk 4]
```diff
@@ -73,19 +74,19 @@ export const IcaAccountHolderI = M.interface('IcaAccountHolder', {
...
     AmountArgShape,
-  ).returns(M.undefined()),
-  withdrawReward: M.callWhen(ChainAddressShape).returns(
-    M.arrayOf(ChainAmountShape),
+  ).returns(VowShape),
+  withdrawReward: M.call(ChainAddressShape).returns(
+    NetworkShape.Vow$(M.arrayOf(ChainAmountShape)),
+  ),
+  withdrawRewards: M.call().returns(
+    NetworkShape.Vow$(M.arrayOf(ChainAmountShape)),
```
Comment (dtribble) [inline packages/orchestration/src/exos/cosmos-orchestration-account.js]: Use of `NetworkShape.Vow$` seems odd, given that there is also `VowShape`.  Can `VowShape` have an optional argument, e.g., `VowShape(NetworkShape...`.  That's pretty clunky, but then there's just one way to express.

File: packages/orchestration/src/exos/cosmos-orchestration-account.js [hunk 5]
```diff
@@ -16,9 +16,10 @@ import {
 } from '@agoric/cosmic-proto/cosmos/staking/v1beta1/tx.js';
 import { Any } from '@agoric/cosmic-proto/google/protobuf/any.js';
 import { makeTracer } from '@agoric/internal';
+import { Shape as NetworkShape } from '@agoric/network';
```
Comment (dtribble) [inline packages/orchestration/src/exos/cosmos-orchestration-account.js:19]: Having to rename `Shape` seems bad.  `network` should not export just `Shape`.

File: packages/orchestration/src/exos/cosmos-orchestration-account.js (hunk 4 above)
Comment (dtribble) [reply packages/orchestration/src/exos/cosmos-orchestration-account.js]: Is there any doc on what `Vow$` is?

File: packages/orchestration/src/exos/cosmos-orchestration-account.js [hunk 6]
```diff
@@ -129,7 +130,7 @@ export const prepareCosmosOrchestrationAccountKit = (
       undelegateWatcher: M.interface('undelegateWatcher', {
         onFulfilled: M.call(M.string())
           .optional(M.arrayOf(M.undefined())) // empty context
-          .returns(M.promise()),
+          .returns(NetworkShape.Vow$(M.promise())),
```
Comment (dtribble) [inline packages/orchestration/src/exos/cosmos-orchestration-account.js]: I cannot parse this pattern at all.

File: packages/orchestration/src/exos/cosmos-orchestration-account.js [hunk 7]
```diff
@@ -138,20 +139,18 @@ export const prepareCosmosOrchestrationAccountKit = (
       }),
       holder: IcaAccountHolderI,
       invitationMakers: M.interface('invitationMakers', {
-        Delegate: M.callWhen(ChainAddressShape, AmountArgShape).returns(
-          InvitationShape,
+        Delegate: M.call(ChainAddressShape, AmountArgShape).returns(
+          M.promise(),
```
Comment (dtribble) [inline packages/orchestration/src/exos/cosmos-orchestration-account.js:143]: shouldn't this be explicitly a promise for `void`?

Comment (dtribble) [reply packages/orchestration/src/exos/cosmos-orchestration-account.js:143]: Does the promise pattern include Vows? Otherwise shouldn't this be a Vow?

File: packages/orchestration/src/exos/cosmos-orchestration-account.js [hunk 8]
```diff
@@ -350,30 +349,32 @@ export const prepareCosmosOrchestrationAccountKit = (
...
+        deposit(payment) {
           trace('deposit', payment);
           console.error(
             'FIXME deposit noop until https://github.com/Agoric/agoric-sdk/issues/9193',
           );
         },
-        async getBalances() {
-          throw Error('not yet implemented');
+        getBalances() {
+          return asVow(() => Fail`not yet implemented`);
```
Comment (dtribble) [inline packages/orchestration/src/exos/cosmos-orchestration-account.js:378]: This should have a reference to the associated bug to implement it, and probably TODO in the error message?

File: packages/orchestration/src/exos/cosmos-orchestration-account.js [hunk 9]
```diff
@@ -382,110 +383,116 @@ export const prepareCosmosOrchestrationAccountKit = (
...
+        withdrawReward(validator) {
+          return asVow(() => {
+            trace('withdrawReward', validator);
+            const { helper } = this.facets;
+            const { chainAddress } = this.state;
+            const msg = MsgWithdrawDelegatorReward.toProtoMsg({
+              delegatorAddress: chainAddress.address,
+              validatorAddress: validator.address,
+            });
+            const account = helper.owned();
```
Comment (dtribble) [inline packages/orchestration/src/exos/cosmos-orchestration-account.js:424]: Am I correct that the `owned` call is the potential failure that needs to be turned into a failed vow? Shoudl that be renamed `tryOwned`?

File: packages/orchestration/src/exos/local-chain-facade.js [hunk 10]
```diff
@@ -1,77 +1,111 @@
...
   );
+harden(prepareLocalChainFacadeKit);
+
+/**
+ * @param {Zone} zone
+ * @param {LocalChainFacadePowers} powers
+ */
+export const prepareLocalChainFacade = (zone, powers) => {
+  const makeLocalChainFacadeKit = prepareLocalChainFacadeKit(zone, powers);
+  return pickFacet(makeLocalChainFacadeKit, 'public');
```
Comment (dtribble) [inline packages/orchestration/src/exos/local-chain-facade.js:109]: This is such a common pattern that we should make better support for it in boilerplate (not in this PR!)

File: packages/smart-wallet/src/offerWatcher.js [hunk 11]
```diff
@@ -37,9 +38,9 @@ import { UNPUBLISHED_RESULT } from './offers.js';
  * @param {UserSeat} seat
  */
 const watchForOfferResult = ({ resultWatcher }, seat) => {
+  // TODO NOT FOR PRODUCTION. See #9308
   const p = E(seat).getOfferResult();
-  watchPromise(p, resultWatcher, seat);
-  return p;
+  return heapVowTools.watch(p, resultWatcher, seat);
```
Comment (dtribble) [inline packages/smart-wallet/src/offerWatcher.js]: This may need `heapVowTools.when` on the front to return a promise rather than a vow.

File: packages/orchestration/test/examples/stake-bld.contract.test.ts [hunk 12]
```diff
@@ -97,18 +97,34 @@ test('makeStakeBldInvitation', async t => {
...
+  {
+    const delegateInv = await E(invitationMakers).Delegate(
+      'agoric1validator1',
+      {
+        brand: bld.brand,
+        value: 1_000_000_000n,
+      },
+    );
+    const delegateOffer = await E(zoe).offer(delegateInv);
+    const res = await E(delegateOffer).getOfferResult();
```
Comment (dtribble) [inline packages/orchestration/test/examples/stake-bld.contract.test.ts:109]: This might now be a `Vow`?

Comment (dtribble) [reply packages/orchestration/test/examples/stake-bld.contract.test.ts:109]: Ah. E is `heapVowE`, so a non-issue

File: packages/orchestration/test/examples/stake-bld.contract.test.ts [hunk 13]
```diff
@@ -97,18 +97,34 @@ test('makeStakeBldInvitation', async t => {
...
+    const delegateInv = await E(invitationMakers).Delegate(
+      'agoric1validator1',
+      {
+        brand: bld.brand,
+        value: 504n,
+      },
+    );
+    const delegateOffer = await E(zoe).offer(delegateInv);
+    await t.throwsAsync(E(delegateOffer).getOfferResult(), {
+      message: 'simualted packet timeout',
```
Comment (dtribble) [inline packages/orchestration/test/examples/stake-bld.contract.test.ts]: "simulated".

File: packages/orchestration/test/examples/stake-atom.contract.test.ts [hunk 14]
```diff
@@ -62,9 +62,6 @@ test('makeAccount, getAddress, getBalances, getBalance', async t => {
   // t.regex(address.address, /cosmos1/);
   t.like(chainAddress, { chainId: 'cosmoshub-4', addressEncoding: 'bech32' });
 
-  t.log('deposit 100 bld to account');
-  await E(account).deposit(await utils.pourPayment(ist.units(100)));
-
   await t.throwsAsync(E(account).getBalances(), {
     message: 'not yet implemented',
```
Comment (dtribble) [inline packages/orchestration/test/examples/stake-atom.contract.test.ts:66]: It seems to me like this should instead be in a `test.failing`.  It's not that the correct behavior is to throw. It's that we know it throws right now due to being unimplemented.

File: packages/orchestration/src/service.js [hunk 15]
```diff
@@ -192,49 +192,44 @@ const prepareOrchestrationKit = (
...
+            {
+              chainId,
+              remoteConnAddr,
+            },
           );
         },
         /**
          * @param {IBCConnectionID} controllerConnectionId
-         * @returns {Promise<ICQConnection>}
+         * @returns {Vow<ICQConnection> | ICQConnection}
```
Comment (dtribble) [inline packages/orchestration/src/service.js:214]: Do we need a reusable pattern for `Vow<T> | T`?  Are there other places where we should have that rather than simply `Vow<T>`?

File: packages/orchestration/src/exos/orchestrator.js [hunk 16]
```diff
@@ -111,27 +112,23 @@ export const prepareOrchestratorKit = (
         },
       },
       orchestrator: {
-        /** @type {Orchestrator['getChain']} */
+        /** @type {PromiseToVow<Orchestrator['getChain']>} */
```
Comment (dtribble) [inline packages/orchestration/src/exos/orchestrator.js:116]: Too much more and we are gonna see renamed to `P2V` :)

File: packages/orchestration/src/exos/orchestrator.js [hunk 17]
```diff
@@ -111,27 +112,23 @@ export const prepareOrchestratorKit = (
...
         getChain(name) {
           if (name === 'agoric') {
-            // XXX when() until membrane
-            return when(
-              watch(
-                chainHub.getChainInfo('agoric'),
-                this.facets.makeLocalChainFacadeWatcher,
-              ),
+            // @ts-expect-error Type 'Vow<Voidless>' is not assignable to type 'Vow<Chain<any>>'.
```
Comment (dtribble) [inline packages/orchestration/src/exos/orchestrator.js:120]: Should there be a TODO/bug to fix these types?

File: packages/boot/test/bootstrapTests/lca.test.ts [hunk 18]
```diff
@@ -89,4 +89,26 @@ test.serial('stakeBld', async t => {
       },
     },
   });
+  t.like(wd.getLatestUpdateRecord(), {
+    status: { id: 'request-delegate', numWantsSatisfied: 1 },
+  });
+
+  await t.throwsAsync(
+    wd.executeOffer({
```
Comment (dtribble) [inline packages/boot/test/bootstrapTests/lca.test.ts:97]: what is the expected error and why?  Seems like somethign we shoudl check the details of, given that there are so many layers that could fail

File: packages/orchestration/src/exos/local-orchestration-account.js [hunk 19]
```diff
@@ -265,32 +263,48 @@ export const prepareLocalOrchestrationAccountKit = (
...
           const [brand, denom] =
             typeof denomArg === 'string'
               ? [/** @type {any} */ (null), denomArg]
               : [denomArg, 'FIXME'];
 
-          const natAmount = await E.when(
+          return watch(
             E(this.state.account).getBalance(brand),
+            this.facets.getBalanceWatcher,
+            { denom },
```
Comment (dtribble) [inline packages/orchestration/src/exos/local-orchestration-account.js]: any particular reason why you pass `denom` in an embedded record rather than as a sole argument? Consistency?  I check reflexively because it is more allocations :)

Comment (dtribble) [reply packages/orchestration/src/exos/local-orchestration-account.js]: but perhaps clearer

File: packages/orchestration/src/exos/local-orchestration-account.js [hunk 20]
```diff
@@ -265,32 +263,48 @@ export const prepareLocalOrchestrationAccountKit = (
...
           );
-          return harden({ denom, value: natAmount.value });
         },
         getBalances() {
-          throw new Error('not yet implemented');
+          return asVow(() => Fail`not yet implemented`);
```
Comment (dtribble) [inline packages/orchestration/src/exos/local-orchestration-account.js:306]: Isn't there a helper already (as with `Promise.reject`) for making and already-rejected vow?

**Review by dtribble — APPROVED** [review]
Excellent work. There was one or two issues that I think really need fixing (and some comments here that should be for later work).

File: packages/boot/test/bootstrapTests/lca.test.ts (hunk 18 above)
Comment (0xpatrickdev) [reply packages/boot/test/bootstrapTests/lca.test.ts:97]: The expected error was commented out at the end of the function (this is the typical place to assert error message values in an ava test).

Inspection of the error message is disabled due to this: https://github.com/Agoric/agoric-sdk/pull/9591/files#r1655705785 - still not quite sure the past way to propagate an error through this bridge device.

The test still serves value since the rejection is properly propagated - just the message is missing.

I have confidence about the layering via tests in `bootstrapTests/orchestration.test.ts`, which show some "ABCI code: 5: error handling packet: see events for details", as well as unit tests for local-orch-acct.

File: packages/orchestration/src/exos/cosmos-orchestration-account.js (hunk 4 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/exos/cosmos-orchestration-account.js]: I created this ticket to clean up interface guards for watchers: https://github.com/Agoric/agoric-sdk/issues/9611

TLDR; agree we should be importing something less verbose from `@agoric/vow`. We also might just go with `VowShape` instead of `Vow$` since the unwrapped shape will likely never be reached. I can't find a link atm but this topic came up in a recent PR review cc @turadg

File: packages/orchestration/src/exos/cosmos-orchestration-account.js (hunk 7 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/exos/cosmos-orchestration-account.js:143]: For the `invitationMakers` interfaces, the return promise unwraps to an `InvitationShape`, not a `VowShape`. The offerHandler of the Invitation is what needs to be a Vow.

File: packages/orchestration/src/exos/cosmos-orchestration-account.js (hunk 8 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/exos/cosmos-orchestration-account.js:378]: Good idea, done: https://github.com/Agoric/agoric-sdk/issues/9610

File: packages/orchestration/src/exos/cosmos-orchestration-account.js (hunk 9 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/exos/cosmos-orchestration-account.js:424]: Correct, that's the reason for `asVow` here. Regarding naming, I believe this pattern was adopted from `vaultHolder`:
https://github.com/Agoric/agoric-sdk/blob/6bc363b5624bab5fd151ec4889b5f5116f2cf53c/packages/inter-protocol/src/vaultFactory/vaultHolder.js#L63-L69

File: packages/orchestration/src/exos/local-orchestration-account.js (hunk 19 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/exos/local-orchestration-account.js]: Context is a single argument, so I’ve been treating it like an opts bag for consistency/readability.

This is good feedback, though. These changes worth considering as we get closer to completion cc @turadg

File: packages/orchestration/src/exos/local-orchestration-account.js (hunk 20 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/exos/local-orchestration-account.js:306]: Not to my knowledge. I think we could also do :
```js
watch(Promise.reject(Fail`...`))
```
But `asVow` seems concise enough to me here

File: packages/orchestration/src/exos/orchestrator.js (hunk 17 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/exos/orchestrator.js:120]: Yes, I added [cb5f043](https://github.com/Agoric/agoric-sdk/pull/9591/commits/cb5f043976d33e1c922388d29683c4e72ec81077) which clearly spells out the compromises I made on types. I suspect the `PromiseVow` added here doesn't get us all the way with generics needed for `getChain`. 

For the ticket, I am attributing it is part of the remaining work for #9449.

File: packages/orchestration/src/service.js (hunk 15 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/service.js:214]: This was the only one I came across so far, but agree its worth considering

File: packages/orchestration/test/examples/stake-atom.contract.test.ts (hunk 14 above)
Comment (0xpatrickdev) [reply packages/orchestration/test/examples/stake-atom.contract.test.ts:66]: It seems like this was there by mistake. This test is for an OrchestrationAccount, not a LocalChainAccount, and we don't currently have a faucet mechanism in unit tests for remote accounts.

Also, with https://github.com/Agoric/agoric-sdk/issues/9193 I believe we decided only LCA/LOA's will have `.deposit()`.

File: .eslintrc.cjs [hunk 21]
```diff
@@ -46,6 +46,11 @@ const resumable = [
     message:
       'callWhen wraps the function in a promise; instead immediately return a vow',
   },
+  {
+    selector: "Identifier[name='heapVowE']",
+    message:
+      'heapVowE shortens vows to promises; instead use `E` from `@endo/far`',
```
Comment (dckc) [inline .eslintrc.cjs]: should be "instead use E with watch()"?

File: packages/orchestration/src/internal.ts [hunk 22]
```diff
@@ -0,0 +1,5 @@
+import type { Vow } from '@agoric/vow';
+
+export type PromiseToVow<T> = T extends (...args: infer A) => Promise<infer R>
```
Comment (dckc) [inline packages/orchestration/src/internal.ts]: I'd like to see docs for `PromiseToVow` and type tests - at least 1 positive and 1 negative.

The name seems long. Maybe it doesn't need to appear in many places?

File: packages/orchestration/src/exos/local-orchestration-account.js [hunk 23]
```diff
@@ -2,14 +2,17 @@
...
 import { TopicsRecordShape } from '@agoric/zoe/src/contractSupport/index.js';
 import { InvitationShape } from '@agoric/zoe/src/typeGuards.js';
+import { E } from '@endo/far';
 import {
   ChainAddressShape,
   ChainAmountShape,
+  DenomAmountShape,
+  DenomShape,
   IBCTransferOptionsShape,
 } from '../typeGuards.js';
```
Comment (dckc) [inline packages/orchestration/src/exos/local-orchestration-account.js:16]: aside: we call them `ThingShape` but put them in files called `typeGuards`. In E, they were called [Guards](http://wiki.erights.org/wiki/Guard-based_auditing). (likewise [guards in Monte](https://monte.readthedocs.io/en/latest/glossary.html#term-guard-coercion)).

Trying it out just to look at it: `DenomGuard` vs. `DenomShape`. Hm. Not much of an improvement.

But our dev audience might already be familiar with the notion by the name "guard":

> The term is used with specific meaning in [APL](https://en.wikipedia.org/wiki/APL_(programming_language)), [Haskell](https://en.wikipedia.org/wiki/Haskell_(programming_language)), [Clean](https://en.wikipedia.org/wiki/Clean_programming_language), [Erlang](https://en.wikipedia.org/wiki/Erlang_programming_language), [occam](https://en.wikipedia.org/wiki/Occam_(programming_language)), [Promela](https://en.wikipedia.org/wiki/Promela), [OCaml](https://en.wikipedia.org/wiki/OCaml), [Swift](https://en.wikipedia.org/wiki/Swift_(programming_language)),[[2]](https://en.wikipedia.org/wiki/Guard_(computer_science)#cite_note-2) [Python](https://en.wikipedia.org/wiki/Python_(programming_language)) from version 3.10, and [Scala](https://en.wikipedia.org/wiki/Scala_(programming_language)) programming languages.
 -- [Guard \(computer science\) \- Wikipedia](https://en.wikipedia.org/wiki/Guard_(computer_science))

File: packages/orchestration/src/exos/local-orchestration-account.js [hunk 24]
```diff
@@ -47,10 +50,12 @@ const { Fail } = assert;
 const HolderI = M.interface('holder', {
   ...orchestrationAccountMethods,
   getPublicTopics: M.call().returns(TopicsRecordShape),
-  delegate: M.call(M.string(), AmountShape).returns(M.promise()),
-  undelegate: M.call(M.string(), AmountShape).returns(M.promise()),
-  withdraw: M.callWhen(AmountShape).returns(PaymentShape),
-  executeTx: M.callWhen(M.arrayOf(M.record())).returns(M.arrayOf(M.record())),
+  delegate: M.call(M.string(), AmountShape).returns(VowShape),
+  undelegate: M.call(M.string(), AmountShape).returns(VowShape),
+  withdraw: M.call(AmountShape).returns(NetworkShape.Vow$(PaymentShape)),
```
Comment (dckc) [inline packages/orchestration/src/exos/local-orchestration-account.js]: `Network` is a distraction here.

I suggest doing something like `const { Vow$ } = NetworkShape;` above.

File: packages/orchestration/src/exos/local-orchestration-account.js [hunk 25]
```diff
@@ -173,14 +185,18 @@ export const prepareLocalOrchestrationAccountKit = (
       undelegateWatcher: {
         /**
          * @param {[
-         *   TypedJson<'/cosmos.staking.v1beta1.MsgUndelegateResponse'>,
+         *   JsonSafe<
+         *     TypedJson<'/cosmos.staking.v1beta1.MsgUndelegateResponse'>
```
Comment (dckc) [inline packages/orchestration/src/exos/local-orchestration-account.js:180]: note to self: is this distinction also expressed in a test?

File: packages/orchestration/src/exos/local-orchestration-account.js [hunk 26]
```diff
@@ -265,29 +283,45 @@ export const prepareLocalOrchestrationAccountKit = (
        */
       returnVoidWatcher: {
         /**
-         * @param {Record<unknown, unknown>[]} results
+         * @param {any} _result
```
Comment (dckc) [inline packages/orchestration/src/exos/local-orchestration-account.js]: Please let's avoid `any` unless there's a strong reason to use it.
Use `unknown` instead; it's type-safe.

File: packages/orchestration/src/exos/local-orchestration-account.js [hunk 27]
```diff
@@ -265,29 +283,45 @@ export const prepareLocalOrchestrationAccountKit = (
...
-          results.length === 1 ||
-            Fail`expected exactly one result; got ${results}`;
-          trace('Result', results[0]);
+        onFulfilled(_result) {
           return undefined;
         },
       },
+      /**
+       * handles a request for balance from a bank purse and returns the balance
+       * as a Chain Amount
```
Comment (dckc) [inline packages/orchestration/src/exos/local-orchestration-account.js:272]: I was going to suggest writing "returns the balance as a Chain Amount" using JSDoc so the computer can help us maintain it... I see you've got `@returns {DenomAmount}`, so in some sense you already did it, but in fact this proves the point: it should be `DenomAmount`, not "Chain Amount".

I'm not sure this comment is worth the bother. It doesn't actually handle a request for a balance. It provides callbacks for a watch and such.

File: packages/orchestration/src/exos/local-orchestration-account.js (hunk 19 above)
Comment (dckc) [reply packages/orchestration/src/exos/local-orchestration-account.js]: > ... These changes worth considering as we get closer to completion ...

I hope any plans to re-consider code we've written are represented by open issues.

File: packages/orchestration/src/exos/local-orchestration-account.js [hunk 28]
```diff
@@ -307,76 +341,78 @@ export const prepareLocalOrchestrationAccountKit = (
...
-          return when(watch(results, this.facets.extractFirstResultWatcher));
+          return watch(
+            E(lca).executeTx([
+              typedJson('/cosmos.staking.v1beta1.MsgDelegate', {
+                amount,
+                validatorAddress,
+                delegatorAddress: this.state.address.address,
+              }),
+            ]),
+            this.facets.extractFirstResultWatcher,
```
Comment (dckc) [inline packages/orchestration/src/exos/local-orchestration-account.js:339]: Is including `Watcher` in the name worthwhile? It's starting to feel like `extractFirstResultWatcherBehaviorDefinitionCodeSpecification`.

File: packages/orchestration/src/exos/local-orchestration-account.js [hunk 29]
```diff
@@ -385,39 +421,43 @@ export const prepareLocalOrchestrationAccountKit = (
...
+            const transferV = watch(
+              allVows([connectionInfoV, timeoutTimestampV]),
+              this.facets.transferWatcher,
+              { opts, amount, destination },
+            );
+            return watch(transferV, this.facets.returnVoidWatcher);
+          });
         },
-        /** @type {OrchestrationAccount<any>['transferSteps']} */
+        /** @type {PromiseToVow<OrchestrationAccount<any>['transferSteps']>} */
```
Comment (dckc) [inline packages/orchestration/src/exos/local-orchestration-account.js:433]: This usage is a good basis for a type test; that is: in the test, use these names as motivation.

File: packages/orchestration/src/utils/orchestrationAccount.js [hunk 30]
```diff
@@ -1,18 +1,20 @@
...
 
 /** @see {OrchestrationAccountI} */
 export const orchestrationAccountMethods = {
   getAddress: M.call().returns(ChainAddressShape),
-  getBalance: M.callWhen(M.any()).returns(CoinShape),
-  getBalances: M.callWhen().returns(M.arrayOf(CoinShape)),
-  send: M.callWhen(ChainAddressShape, AmountArgShape).returns(M.undefined()),
-  transfer: M.callWhen(AmountArgShape, ChainAddressShape)
+  getBalance: M.call(M.any()).returns(NetworkShape.Vow$(CoinShape)),
+  getBalances: M.call().returns(NetworkShape.Vow$(M.arrayOf(CoinShape))),
```
Comment (dckc) [inline packages/orchestration/src/utils/orchestrationAccount.js]: again, `Network` is a distraction here

File: packages/orchestration/src/exos/local-orchestration-account.js [hunk 31]
```diff
@@ -155,27 +153,28 @@ export const prepareLocalOrchestrationAccountKit = (
...
-        async Delegate(validatorAddress, ertpAmount) {
+        Delegate(validatorAddress, ertpAmount) {
           trace('Delegate', validatorAddress, ertpAmount);
 
-          return zcf.makeInvitation(async seat => {
-            // TODO should it allow delegating more BLD?
+          return zcf.makeInvitation(seat => {
             seat.exit();
-            return this.facets.holder.delegate(validatorAddress, ertpAmount);
+            return watch(
```
Comment (dckc) [inline packages/orchestration/src/exos/local-orchestration-account.js:153]: note to self: is a test for this change included in this PR?

If not, what's our plan for testing it?

Comment (dckc) [reply packages/orchestration/src/exos/local-orchestration-account.js:153]: the 1st line of the commit message exceeds conventional norms

I suggest changing `feat(orchestration):` to `chore:`. The feature that I would expect to see in release notes is local-orchestration-account.

File: packages/orchestration/src/exos/local-chain-facade.js [hunk 32]
```diff
@@ -54,8 +54,8 @@ export const prepareLocalChainFacade = (
       makeAccount() {
         const { localChainInfo } = this.state;
         const lcaP = E(localchain).makeAccount();
-        // FIXME use watch() from vowTools
-        return heapVowTools.watch(allVows([lcaP, E(lcaP).getAddress()]), {
+        // @ts-expect-error E does not understand Vow pipelining
```
Comment (dckc) [inline packages/orchestration/src/exos/local-chain-facade.js]: That seems like a devex issue that our customers may encounter.

IOU an issue. Bonus points if you beat me to it.
cc @Jovonni @mitdralla

**Review by dckc — COMMENTED** [review]
time for a meeting; sharing what I've got so far

File: packages/orchestration/src/exos/local-chain-facade.js (hunk 32 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/exos/local-chain-facade.js]: >  E does not understand Vow pipelining

I thought this was the case, but it doesn't seem to be. I was tricked by the linter at one point and this comment is not in the final revision. I can try to squash it out of the commit, but not sure it's worth the effort

Comment (0xpatrickdev) [reply packages/orchestration/src/exos/local-chain-facade.js]: On further analysis, the original comment does appear to be true. In this particular instance, `E(localchain).makeAccount` is a Promise, not a Vow, so pipelining works.

Here's a test that demonstrates the behavior: [d7bf05d](https://github.com/Agoric/agoric-sdk/pull/9591/commits/d7bf05d5b4ab8d6a69b730713355fd607480a424)

File: packages/orchestration/src/internal.ts (hunk 22 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/internal.ts]: Please see tests added here: [2dcc26f](https://github.com/Agoric/agoric-sdk/pull/9591/commits/2dcc26f29de7c2cbc1506f62293934341fa4dbaf). As always, thanks for careful review. Something was missed in the original implementation.

File: packages/orchestration/src/exos/local-orchestration-account.js (hunk 28 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/exos/local-orchestration-account.js:339]: I agree its verbose. This seems to be the pattern adopted elsewhere, so I ran with it. Technically it's a "watcher handler"

File: packages/orchestration/src/exos/local-orchestration-account.js (hunk 19 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/exos/local-orchestration-account.js]: Not worth a ticket so updated here: [69a6e21](https://github.com/Agoric/agoric-sdk/pull/9591/commits/69a6e2157b8d1223adfd73ff7d8927ea67628e3b)

File: packages/orchestration/src/internal.ts (hunk 22 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/internal.ts]: "Missed" is maybe harsh. It would return `never` if the target was not a Promise. The caller shouldn't be using it in the first place. 

Now, it passes through all values and only transforms Promises to Vows.

File: packages/orchestration/src/exos/cosmos-orchestration-account.js (hunk 7 above)
Comment (dckc) [reply packages/orchestration/src/exos/cosmos-orchestration-account.js:143]: > shouldn't this be explicitly a promise for void?

That's not expressible as a pattern. It's not pass-invariant.

Comment (erights) [reply packages/orchestration/src/exos/cosmos-orchestration-account.js:143]: (Nit: in typescript `void` does not signify `undefined` but rather a kind of don't care. I'll ignore that below.)

In is indeed not expressible as a pattern, but in this case not because of pass invariance. Patterns (as opposed to guards) can only enforce what they can immediately check, and a pattern cannot immediately check what the promise will eventually fulfill to. 

The original, by using an `M.callWhen` method guard with a `returns(InvitationShape)` does enforce that `Delegate` returns a promise for something that matches `InvitationShape`, because it uses the `callWhen` delay to validate the promise's fulfillment before passing it through. Likewise, within an `M.callWhen` you can say `M.await(FooShape)` as a parameter guard. If the argument is a promise, this will enforce that the promise fulfills to something that matches `FooShape` before passing that fulfillment along as the argument to the raw method.
